src/exceptions.rs

```diff
@@ -754,8 +759,30 @@ impl PyUnicodeDecodeError {
         input: &[u8],
         err: std::str::Utf8Error,
     ) -> PyResult<Bound<'py, PyUnicodeDecodeError>> {
-        let pos = err.valid_up_to();
-        PyUnicodeDecodeError::new(py, c"utf-8", input, pos..(pos + 1), c"invalid utf-8")
+        let start = err.valid_up_to();
+        let end = err.error_len().map_or(input.len(), |l| start + l);
+        PyUnicodeDecodeError::new(py, c"utf-8", input, start..end, c"invalid utf-8")
+    }
+
+    /// Create a new [`crate::PyErr`] of this type from a Rust UTF-8 decoding error.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use pyo3::prelude::*;
+    /// use pyo3::exceptions::PyUnicodeDecodeError;
+    ///
+    /// let invalid_utf8 = b"fo\xd8o";
+    /// # #[expect(invalid_from_utf8)]
+    /// let err = std::str::from_utf8(invalid_utf8).expect_err("should be invalid utf8");
+    /// let py_err = PyUnicodeDecodeError::new_err_from_utf8(invalid_utf8, err);
+    /// ```
+    pub fn new_err_from_utf8(bytes: &[u8], err: std::str::Utf8Error) -> crate::PyErr {
+        Utf8ErrorWithBytes {
+            err,
+            bytes: bytes.to_vec(),
```

Comment:
Can we avoid this intermediate allocation? I guess not, because we don't have `py: Python` here.

... is it possibly better to take `'py: Python` as an argument, call `new_utf8` above to construct Python bytes immediately, and then use `PyErr::from_value` inside this function?

Reply:
... or we just return `PyResult<PyErr>` here and let callers decide whether to flatten the other errors away, but the API does feel a bit awkward then.

Reply:
Yeah I didn't really want to return a `PyResult<PyErr>`, in part because it would be an awkward API and in part to keep symmetry between `new_err_from_utf8` and `new_err`.

Adding a `py: Python` argument would also break the symmetry with `new_err`, but if you think that's worth it to avoid the extra allocation then I can make that change